### PR TITLE
EDU-941 Extract publicContext within document model

### PR DESCRIPTION
**Pull request by adrianaluchian**
Closes https://educandu.atlassian.net/browse/EDU-941

### CODE REVIEW

**Comment by ahelmberger** [comment]
The "new" dialog is much cleaner, great work! :heart_eyes:

I found one deviation from the former behavior though, that we have to correct:

When cloning a document, the old implementation resets the four properties that make up the `publicContext` to their default values, while the reworked solution keeps the original values from the source document (at least as long as the cloning strategy dropdown stays unchanged). This is problematic as it:

* prevents non-maintainers from cloning+editing any document with restricted `allowedOpenContributions`
* enables "stealing" of the `verified` status by simply cloning and modifying an already verified document

In other words: in the `resetStates` function we have to disregard the `initialDocumentMetadata.publicContext` values when the dialog was opened in cloning mode.

File: src/components/document-metadata-modal.js [hunk 1]
```diff
@@ -56,47 +67,65 @@ function DocumentMetadataModal({
...
+  );
+
+  const resetStates = useCallback(() => {
+    setTitle(initialDocumentMetadata.title || t('newDocument'));
+    setDescription(initialDocumentMetadata.description || '');
+    setSlug(initialDocumentMetadata.slug || '');
+    setTags(initialDocumentMetadata.tags || []);
+    setTagOptions(composeTagOptions(initialDocumentMetadata.tags));
+    setLanguage(initialDocumentMetadata.language || getDefaultLanguageFromUiLanguage(uiLanguage));
+    setPublicContext(initialDocumentMetadata.publicContext || getDefaultPublicContext());
```
Comment (ahelmberger) [inline src/components/document-metadata-modal.js:121]: It would feel cleaner if we'd `deepClone`d this object (although technically irrelevant). Also, please see my comment at the end of this PR on the cloning case.

File: src/components/document-metadata-modal.yml [hunk 2]
```diff
@@ -7,9 +7,6 @@ cloneDocument:
 editDocument:
   en: Edit document metadata
   de: Dokument-Metadaten bearbeiten
-roomRequired:
-  en: Please specify a room
-  de: Bitte gib einen Raum an
```
Comment (ahelmberger) [inline src/components/document-metadata-modal.yml:12]: Did you delete this validation text on purpose (because the label of the field is more or less the same)?

File: src/domain/schemas/document-schemas.js [hunk 3]
```diff
@@ -15,17 +15,31 @@ export const patchDocSectionsBodySchema = joi.object({
...
+  allowedOpenContribution: joi.string().valid(...Object.values(DOCUMENT_ALLOWED_OPEN_CONTRIBUTION)).required()
+});
+
+export const publicContextConditionalSchema = joi.alternatives().conditional(
+  'roomId',
+  {
+    is: idOrKeySchema,
+    then: null,
+    otherwise: publicContextSchema.required()
+  }
```
Comment (ahelmberger) [inline src/domain/schemas/document-schemas.js:31]: For some reason I would have expected the inverted test:

~~~
{
  is: null,
  then: publicContextSchema.required(),
  otherwise: null
}
~~~

But it's probably a matter of taste. Consider this comment purely informational and non-passionate :smile:

File: src/components/document-metadata-modal.yml (hunk 2 above)
Comment (adrianaluchian) [reply src/components/document-metadata-modal.yml:12]: Initially I even thought that the label was wrong because it was phrased like a validation message 😅 So I reduced them.

File: src/domain/schemas/document-schemas.js (hunk 3 above)
Comment (adrianaluchian) [reply src/domain/schemas/document-schemas.js:31]: This is just temporary, it will make sense when we add the schema for `roomContext`, in which case `is: idOrKeySchema` is the setup that makes more sense.

**Comment by adrianaluchian** [comment]
> The "new" dialog is much cleaner, great work! 😍
> 
> I found one deviation from the former behavior though, that we have to correct:
> 
> When cloning a document, the old implementation resets the four properties that make up the `publicContext` to their default values, while the reworked solution keeps the original values from the source document (at least as long as the cloning strategy dropdown stays unchanged). This is problematic as it:
> 
> * prevents non-maintainers from cloning+editing any document with restricted `allowedOpenContributions`
> * enables "stealing" of the `verified` status by simply cloning and modifying an already verified document
> 
> In other words: in the `resetStates` function we have to disregard the `initialDocumentMetadata.publicContext` values when the dialog was opened in cloning mode.

Very good catch, thank you! I've fixed it.

File: src/components/document-metadata-modal.js (hunk 1 above)
Comment (adrianaluchian) [reply src/components/document-metadata-modal.js:121]: Cleaner maybe no, but definitely safer! Cloned it, thanks.

File: src/domain/schemas/document-schemas.js (hunk 3 above)
Comment (adrianaluchian) [reply src/domain/schemas/document-schemas.js:31]: Actually it's been long enough that I forgot the code. This is a dedicated schema for public context only.. In which case what you suggested is more logical. I am changing this already in the follow-up ticket.

Comment (ahelmberger) [reply src/domain/schemas/document-schemas.js:31]: Poifect!

File: src/components/document-metadata-modal.yml (hunk 2 above)
Comment (ahelmberger) [reply src/components/document-metadata-modal.yml:12]: So, as we agreed, we'll bring back this validation message and rephrase the label to

* eng: "Room"
* ger: "Raum"

Comment (adrianaluchian) [reply src/components/document-metadata-modal.yml]: Aye aye, done!

Comment (ahelmberger) [reply src/components/document-metadata-modal.yml]: Cool, thanks!

**Review by ahelmberger — APPROVED** [review]
